packages/synchronize/lib/src/sync.dart

```diff
@@ -0,0 +1,246 @@
+import 'package:synchronize/src/action.dart';
+import 'package:synchronize/src/conflict.dart';
+import 'package:synchronize/src/journal.dart';
+import 'package:synchronize/src/journal_entry.dart';
+import 'package:synchronize/src/object.dart';
+import 'package:synchronize/src/sources.dart';
+
+/// Sync between two [SyncSources]s.
+///
+/// This implementation follows https://unterwaditzer.net/2016/sync-algorithm.html in a generic and abstract way
+/// and should work for any two kinds of sources and objects.
+Future<List<SyncConflict<T1, T2>>> sync<T1, T2>(
+  final SyncSources<T1, T2> sources,
+  final SyncJournal journal, {
+  final Map<String, SyncConflictSolution> conflictSolutions = const {},
```

Comment:
I was more thinking of making it nullable as I think const collections can be dangerous.
I can see me 100% calling conflictSolutions.add down the road and wondering why it's not working.

I briefly tested it this morning and it should be possible with null.

Reply:
Ah true, that could be a problem indeed